File: scripts/fetch_images.py
"""Fetch Wikipedia page images for all painters in painters.json
and download them to a local images/ directory.
Usage:
    python3 scripts/fetch_images.py
"""
import json
import time
import os
import re
import urllib.request
import urllib.parse
import urllib.error

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PAINTERS_FILE = os.path.join(BASE_DIR, 'painters.json')
IMAGES_DIR = os.path.join(BASE_DIR, 'images')
BATCH_SIZE = 50
API_URL = 'https://en.wikipedia.org/w/api.php'


def sanitize_filename(name):
    """Turn a painter name into a safe filename."""
    s = name.lower()
    s = re.sub(r'[^a-z0-9]+', '_', s)
    s = s.strip('_')
    return s[:80]


def download_image(url, filepath):
    """Download image from url to filepath. Return True on success."""
    try:
        req = urllib.request.Request(url, headers={
            'User-Agent': 'PainterList/1.0 (painter catalog; contact@example.com)'
        })
        with urllib.request.urlopen(req, timeout=30) as resp:
            with open(filepath, 'wb') as f:
                f.write(resp.read())
        return True
    except Exception as e:
        print(f"    Download failed: {e}")
        return False


def main():
    os.makedirs(IMAGES_DIR, exist_ok=True)

    with open(PAINTERS_FILE) as f:
        painters = json.load(f)

    remaining = [p for p in painters if 'image_url' not in p or 'image_path' not in p]
    if not remaining:
        print("All painters already processed. Nothing to do.")
        return

    total = len(painters)
    print(f"Processing {total} painters...")

    for start in range(0, total, BATCH_SIZE):
        batch = painters[start:start + BATCH_SIZE]
        titles = [p['name'] for p in batch]

        # --- Step 1: Fetch image URLs from Wikipedia API ---
        params = urllib.parse.urlencode({
            'action': 'query',
            'prop': 'pageimages',
            'pithumbsize': 200,
            'titles': '|'.join(titles),
            'redirects': 1,
            'format': 'json',
            'origin': '*'
        })
        url = f'{API_URL}?{params}'

        try:
            with urllib.request.urlopen(url, timeout=30) as resp:
                data = json.loads(resp.read())

            redirect_map = {}
            for r in data.get('query', {}).get('redirects', []):
                redirect_map[r['from']] = r['to']

            title_thumb = {}
            for page_id, page_data in data.get('query', {}).get('pages', {}).items():
                if page_id == '-1':
                    continue
                title = page_data.get('title', '')
                thumb = page_data.get('thumbnail')
                if thumb:
                    title_thumb[title] = thumb['source']

            for painter in batch:
                name = painter['name']
                img_url = title_thumb.get(name) or title_thumb.get(redirect_map.get(name))
                if img_url:
                    painter['image_url'] = img_url

        except Exception as e:
            print(f"  Batch API call failed: {e}")

        # --- Step 2: Download images for this batch ---
        for painter in batch:
            img_url = painter.get('image_url')
            if not img_url:
                continue

            # Determine file extension from URL
            ext = os.path.splitext(urllib.parse.urlparse(img_url).path)[1]
            if not ext or ext.lower() not in ('.jpg', '.jpeg', '.png', '.gif', '.webp'):
                ext = '.jpg'

            safe_name = sanitize_filename(painter['name'])
            filename = f"{painter['id']:04d}_{safe_name}{ext}"
            filepath = os.path.join(IMAGES_DIR, filename)

            if os.path.exists(filepath):
                painter['image_path'] = filepath
                continue

            if download_image(img_url, filepath):
                painter['image_path'] = filepath
                print(f"    Downloaded: {filename}")
            else:
                # Remove invalid URL so retry re-fetches
                painter.pop('image_url', None)

        time.sleep(0.5)

        # --- Progress ---
        with_url = sum(1 for p in painters if 'image_url' in p)
        with_file = sum(1 for p in painters if 'image_path' in p)
        done = min(start + BATCH_SIZE, total)
        print(f"  {done}/{total} — {with_url} URLs, {with_file} files downloaded")

    # Save updated JSON
    with open(PAINTERS_FILE, 'w') as f:
        json.dump(painters, f, ensure_ascii=False, indent=2)

    with_url = sum(1 for p in painters if 'image_url' in p)
    with_file = sum(1 for p in painters if 'image_path' in p)
    print(f"\nDone. {with_url}/{total} painters with image_url, {with_file} files in {IMAGES_DIR}/")


if __name__ == '__main__':
    main()

File: scripts/fetch_images.mjs
/**
 * Fetch Wikipedia page images for all painters in painters.json
 * and download them to a local images/ directory.
 *
 * Usage:
 *   node scripts/fetch_images.mjs
 */
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { createWriteStream } from 'node:fs';
import { get } from 'node:https';
import { resolve, dirname, extname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const BASE_DIR = resolve(dirname(__filename), '..');
const PAINTERS_FILE = join(BASE_DIR, 'painters.json');
const IMAGES_DIR = join(BASE_DIR, 'images');
const BATCH_SIZE = 50;
const API_URL = 'https://en.wikipedia.org/w/api.php';

function sanitizeFilename(name) {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_|_$/g, '')
    .slice(0, 80);
}

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    get(url, { headers: { 'User-Agent': 'PainterList/1.0' } }, (res) => {
      if (res.statusCode < 200 || res.statusCode >= 300) {
        reject(new Error(`HTTP ${res.statusCode}`));
        res.resume();
        return;
      }
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end', () => resolve(Buffer.concat(chunks)));
    }).on('error', reject);
  });
}

async function downloadImage(url, filepath) {
  try {
    const data = await httpsGet(url);
    return new Promise((resolve) => {
      const ws = createWriteStream(filepath);
      ws.write(data);
      ws.end();
      ws.on('finish', () => resolve(true));
      ws.on('error', () => resolve(false));
    });
  } catch {
    return false;
  }
}

function getExtension(url) {
  const pathname = new URL(url).pathname;
  const ext = extname(pathname).toLowerCase();
  return ['.jpg', '.jpeg', '.png', '.gif', '.webp'].includes(ext) ? ext : '.jpg';
}

async function main() {
  await mkdir(IMAGES_DIR, { recursive: true });

  const raw = await readFile(PAINTERS_FILE, 'utf-8');
  const painters = JSON.parse(raw);

  const hasAll = painters.every((p) => p.image_path);
  if (hasAll) {
    console.log('All painters already have image_path. Nothing to do.');
    return;
  }

  const total = painters.length;
  console.log(`Processing ${total} painters...`);

  for (let start = 0; start < total; start += BATCH_SIZE) {
    const batch = painters.slice(start, start + BATCH_SIZE);
    const titles = batch.map((p) => p.name);

    // Step 1: Fetch image URLs from Wikipedia API
    const params = new URLSearchParams({
      action: 'query',
      prop: 'pageimages',
      pithumbsize: '200',
      titles: titles.join('|'),
      redirects: '1',
      format: 'json',
      origin: '*',
    });
    const apiUrl = `${API_URL}?${params}`;

    try {
      const buf = await httpsGet(apiUrl);
      const data = JSON.parse(buf.toString());

      const redirectMap = {};
      for (const r of data.query?.redirects ?? []) {
        redirectMap[r.from] = r.to;
      }

      const titleThumb = {};
      for (const pageData of Object.values(data.query?.pages ?? {})) {
        if (pageData.title && pageData.thumbnail?.source) {
          titleThumb[pageData.title] = pageData.thumbnail.source;
        }
      }

      for (const painter of batch) {
        const url =
          titleThumb[painter.name] ?? titleThumb[redirectMap[painter.name]];
        if (url) painter.image_url = url;
      }
    } catch (err) {
      console.log(`  Batch API call failed: ${err.message}`);
    }

    // Step 2: Download images
    for (const painter of batch) {
      const imgUrl = painter.image_url;
      if (!imgUrl) continue;

      const ext = getExtension(imgUrl);
      const safeName = sanitizeFilename(painter.name);
      const filename = `${String(painter.id).padStart(4, '0')}_${safeName}${ext}`;
      const filepath = join(IMAGES_DIR, filename);

      if (existsSync(filepath)) {
        painter.image_path = filepath;
        continue;
      }

      const ok = await downloadImage(imgUrl, filepath);
      if (ok) {
        painter.image_path = filepath;
        console.log(`    Downloaded: ${filename}`);
      } else {
        delete painter.image_url;
      }
    }

    await new Promise((r) => setTimeout(r, 500));

    const withUrl = painters.filter((p) => p.image_url).length;
    const withFile = painters.filter((p) => p.image_path).length;
    const done = Math.min(start + BATCH_SIZE, total);
    console.log(`  ${done}/${total} — ${withUrl} URLs, ${withFile} files`);
  }

  await writeFile(PAINTERS_FILE, JSON.stringify(painters, null, 2), 'utf-8');

  const withUrl = painters.filter((p) => p.image_url).length;
  const withFile = painters.filter((p) => p.image_path).length;
  console.log(`\nDone. ${withUrl}/${total} painters with image_url, ${withFile} files in ${IMAGES_DIR}/`);
}

main().catch(console.error);
